[quant] fix int16 quantization scale in conv weight

terrychenism opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74665

Summary:
fix int16 quantization scale in conv weight
before pr, ref_module.conv.get_quantized_weight() in qint32 dtype will get scale = 1.0.
fixed by add qint32 support in set/get weight function for reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_int16_reference_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35106497](https://our.internmc.facebook.com/intern/diff/D35106497)

terrychenism committed:
[quant] fix int16 quantization scale in conv weight

Summary:
fix int16 quantization scale in conv weight

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_int16_reference_module

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

vkuzo commented:
> fix int16 quantization scale in conv weight

In the PR summary, could we provide some context on what specifically was broken, and how you are fixing it?

vkuzo commented on test/quantization/eager/test_quantize_eager_ptq.py:
```diff
@@ -199,6 +199,61 @@ def test_linear(self):
+        data = torch.randn(*input_size, dtype=torch.float)
+
+        original_ref_m = RefM()
+        torch.quantization.engine = "qnnpack"
```
note: not all test environments will support qnnpack.  You can check for this with `"qnnpack" in torch.testing._internal.common_quantized.supported_qengines`, and skip the test if this is false

jerryzh168 commented on torch/nn/quantized/_reference/modules/utils.py:
```diff
@@ -98,11 +98,11 @@ def _quantize_weight(
 
     if weight_qscheme == torch.per_tensor_affine:
-        if weight_dtype in [torch.quint8, torch.qint8, torch.qint32]:
+        if weight_dtype in [torch.quint8, torch.qint8, torch.qint32, torch.qint32]:
```
nit: torch.qint32 already included before for this one

vkuzo commented on torch/nn/quantized/_reference/modules/utils.py:
```diff
@@ -16,7 +16,7 @@ def _init_weight_qparams(self, weight_qparams, device):
             torch.per_channel_affine_float_qparams], \
             Exception(f"qscheme: {self.weight_qscheme} is not support in reference quantized {self._get_name()}")
-        if self.weight_dtype in [torch.quint8, torch.qint8, torch.quint4x2]:
+        if self.weight_dtype in [torch.quint8, torch.qint8, torch.quint4x2, torch.qint32]:
```
can we add a warning on L37 in the else branch to enable the user to find this easier next time?

jerryzh168 replied:
I think we can add an assert, the `else` branch should not be called

terrychenism replied:
I added 'if-else' back due to one of fx test cases needs torch.float, so it will call 'else'

terrychenism replied:
TestQuantizeFx.test_dynamic_with_fusion_multiple_uses

terrychenism committed:
Update on "[quant] fix int16 quantization scale in conv weight"


Summary:
fix int16 quantization scale in conv weight
before pr, ref_module.conv.get_quantized_weight() in qint32 dtype will get scale = 1.0.
fixed by add qint32 support in set/get weight function for reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_int16_reference_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35106497](https://our.internmc.facebook.com/intern/diff/D35106497)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

jerryzh168 commented on test/quantization/eager/test_quantize_eager_ptq.py:
```diff
@@ -199,6 +199,64 @@ def test_linear(self):
+        ref_m(torch.randn(*input_size, dtype=torch.float))
+
+        ref_m = convert(ref_m, is_reference=True)
+        self.assertNotEqual(ref_m.conv.weight_scale, 1.0)
```
hmmm, I think a better test is to initialize a weight observer here and calculate the scale and check it is the same as ref_m.conv.weight_scale

jerryzh168 commented on torch/nn/quantized/_reference/modules/utils.py:
```diff
@@ -35,6 +39,7 @@ def _init_weight_qparams(self, weight_qparams, device):
                 self.register_buffer(
                     "weight_axis", torch.tensor(0, dtype=torch.int, device=device))
         else:
+            logger.warning("unexpected supported weight dtype: {}".format(self.weight_dtype))
```
I think we can change this to assert, this code is not expected to be executed

jerryzh168 commented on torch/nn/quantized/_reference/modules/utils.py:
```diff
@@ -16,7 +20,7 @@ def _init_weight_qparams(self, weight_qparams, device):
             torch.per_channel_affine_float_qparams], \
             Exception(f"qscheme: {self.weight_qscheme} is not support in reference quantized {self._get_name()}")
-        if self.weight_dtype in [torch.quint8, torch.qint8, torch.quint4x2]:
+        if self.weight_dtype in [torch.quint8, torch.qint8, torch.quint4x2, torch.qint32]:
```
actually I feel we can just change this `if` to an assert and remove the else branch..

jerryzh168 reviewed (approved):
looks good overall, had two inline comments, please address them before landing

terrychenism committed:
Update on "[quant] fix int16 quantization scale in conv weight"


Summary:
fix int16 quantization scale in conv weight
before pr, ref_module.conv.get_quantized_weight() in qint32 dtype will get scale = 1.0.
fixed by add qint32 support in set/get weight function for reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_int16_reference_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35106497](https://our.internmc.facebook.com/intern/diff/D35106497)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant] fix int16 quantization scale in conv weight"


Summary:
fix int16 quantization scale in conv weight
before pr, ref_module.conv.get_quantized_weight() in qint32 dtype will get scale = 1.0.
fixed by add qint32 support in set/get weight function for reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_int16_reference_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35106497](https://our.internmc.facebook.com/intern/diff/D35106497)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

terrychenism committed:
Update on "[quant] fix int16 quantization scale in conv weight"


Summary:
fix int16 quantization scale in conv weight
before pr, ref_module.conv.get_quantized_weight() in qint32 dtype will get scale = 1.0.
fixed by add qint32 support in set/get weight function for reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_int16_reference_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35106497](https://our.internmc.facebook.com/intern/diff/D35106497)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

terrychenism committed:
Update on "[quant] fix int16 quantization scale in conv weight"


Summary:
fix int16 quantization scale in conv weight
before pr, ref_module.conv.get_quantized_weight() in qint32 dtype will get scale = 1.0.
fixed by add qint32 support in set/get weight function for reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_int16_reference_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35106497](https://our.internmc.facebook.com/intern/diff/D35106497)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

terrychenism committed:
Update on "[quant] fix int16 quantization scale in conv weight"


Summary:
fix int16 quantization scale in conv weight
before pr, ref_module.conv.get_quantized_weight() in qint32 dtype will get scale = 1.0.
fixed by add qint32 support in set/get weight function for reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_int16_reference_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35106497](https://our.internmc.facebook.com/intern/diff/D35106497)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

terrychenism committed:
Update on "[quant] fix int16 quantization scale in conv weight"


Summary:
fix int16 quantization scale in conv weight
before pr, ref_module.conv.get_quantized_weight() in qint32 dtype will get scale = 1.0.
fixed by add qint32 support in set/get weight function for reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_int16_reference_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35106497](https://our.internmc.facebook.com/intern/diff/D35106497)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

terrychenism committed:
Update on "[quant] fix int16 quantization scale in conv weight"


Summary:
fix int16 quantization scale in conv weight
before pr, ref_module.conv.get_quantized_weight() in qint32 dtype will get scale = 1.0.
fixed by add qint32 support in set/get weight function for reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_int16_reference_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35106497](https://our.internmc.facebook.com/intern/diff/D35106497)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant] fix int16 quantization scale in conv weight"


Summary:
fix int16 quantization scale in conv weight
before pr, ref_module.conv.get_quantized_weight() in qint32 dtype will get scale = 1.0.
fixed by add qint32 support in set/get weight function for reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_int16_reference_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35106497](https://our.internmc.facebook.com/intern/diff/D35106497)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).

terrychenism committed:
Update on "[quant] fix int16 quantization scale in conv weight"


Summary:
fix int16 quantization scale in conv weight
before pr, ref_module.conv.get_quantized_weight() in qint32 dtype will get scale = 1.0.
fixed by add qint32 support in set/get weight function for reference module

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_int16_reference_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35106497](https://our.internmc.facebook.com/intern/diff/D35106497)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35106497).
